smartsim/_core/launcher/dragon/pqueue.py

```diff
@@ -0,0 +1,464 @@
+# import collections
+import enum
+import heapq
+import threading
+import typing as t
+
+from smartsim.error.errors import SmartSimError
+from smartsim.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class Node(t.Protocol):
+    """Base Node API required to support the NodePrioritizer"""
+
+    @property
+    def hostname(self) -> str:
+        """The hostname of the node"""
+
+    @property
+    def num_cpus(self) -> int:
+        """The number of CPUs in the node"""
+
+    @property
+    def num_gpus(self) -> int:
+        """The number of GPUs in the node"""
+
+
+class NodeReferenceCount(t.Protocol):
+    """Contains details pertaining to references to a node"""
+
+    @property
+    def hostname(self) -> str:
+        """The hostname of the node"""
+
+    @property
+    def num_refs(self) -> int:
+        """The number of jobs assigned to the node"""
+
+
+class _TrackedNode:
+    """Node API required to have support in the NodePrioritizer"""
+
+    def __init__(self, node: Node) -> None:
+        self._node = node
+        """The node being tracked"""
+        self._num_refs = 0
+        """The number of references to the tracked node"""
+        self._assigned_tasks: t.Set[str] = set()
+        """The unique identifiers of processes using this node"""
+        self._is_dirty = False
+        """Flag indicating that tracking information has been modified"""
+
+    @property
+    def hostname(self) -> str:
+        """Returns the hostname of the node"""
+        return self._node.hostname
+
+    @property
+    def num_cpus(self) -> int:
+        """Returns the number of CPUs in the node"""
+        return self._node.num_cpus
+
+    @property
+    def num_gpus(self) -> int:
+        """Returns the number of GPUs attached to the node"""
+        return self._node.num_gpus
+
+    @property
+    def num_refs(self) -> int:
+        """Returns the number of processes currently running on the node"""
+        return self._num_refs
+
+    @property
+    def is_assigned(self) -> bool:
+        """Returns `True` if no references are currently counted, `False` otherwise"""
+        return self._num_refs > 0
+
+    @property
+    def assigned_tasks(self) -> t.Set[str]:
+        """Returns the set of unique IDs for currently running processes"""
+        return self._assigned_tasks
+
+    @property
+    def is_dirty(self) -> bool:
+        """Returns a flag indicating if the reference counter has changed. `True`
+        if references have been added or removed, `False` otherwise."""
+        return self._is_dirty
+
+    def clean(self) -> None:
+        """Marks the node as unmodified"""
+        self._is_dirty = False
+
+    def add(
+        self,
+        tracking_id: t.Optional[str] = None,
+    ) -> None:
+        """Update the node to indicate the addition of a process that must be
+        reference counted.
+
+        :param tracking_id: a unique task identifier executing on the node
+        to add"""
```

Comment:
Should we add a ``:raises:`` comment for the ``ValueError``?

Reply:
fixed